tiledb/sm/query/readers/ordered_dim_label_reader.cc

```diff
@@ -0,0 +1,1083 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class OrderedDimLabelReader.
+ */
+
+#include "tiledb/common/logger.h"
+
+#include "tiledb/sm/array/array.h"
+#include "tiledb/sm/array_schema/array_schema.h"
+#include "tiledb/sm/array_schema/dimension.h"
+#include "tiledb/sm/fragment/fragment_metadata.h"
+#include "tiledb/sm/misc/parallel_functions.h"
+#include "tiledb/sm/misc/utils.h"
+#include "tiledb/sm/query/legacy/cell_slab_iter.h"
+#include "tiledb/sm/query/query_macros.h"
+#include "tiledb/sm/query/readers/ordered_dim_label_reader.h"
+#include "tiledb/sm/query/readers/result_tile.h"
+#include "tiledb/sm/stats/global_stats.h"
+#include "tiledb/sm/storage_manager/storage_manager.h"
+#include "tiledb/sm/subarray/subarray.h"
+
+#include <numeric>
+
+using namespace tiledb;
+using namespace tiledb::common;
+using namespace tiledb::sm::stats;
+
+namespace tiledb {
+namespace sm {
+
+class OrderedDimLabelReaderStatusException : public StatusException {
+ public:
+  explicit OrderedDimLabelReaderStatusException(const std::string& message)
+      : StatusException("OrderedDimLabelReader", message) {
+  }
+};
+
+/* ****************************** */
+/*          CONSTRUCTORS          */
+/* ****************************** */
+
+OrderedDimLabelReader::OrderedDimLabelReader(
+    stats::Stats* stats,
+    shared_ptr<Logger> logger,
+    StorageManager* storage_manager,
+    Array* array,
+    Config& config,
+    std::unordered_map<std::string, QueryBuffer>& buffers,
+    Subarray& subarray,
+    Layout layout,
+    QueryCondition& condition,
+    bool increasing_labels,
+    bool skip_checks_serialization)
+    : ReaderBase(
+          stats,
+          logger->clone("OrderedDimLabelReader", ++logger_id_),
+          storage_manager,
+          array,
+          config,
+          buffers,
+          subarray,
+          layout,
+          condition)
+    , ranges_(
+          subarray.get_attribute_ranges(array_schema_.attributes()[0]->name()))
+    , label_name_(array_schema_.attributes()[0]->name())
+    , label_type_(array_schema_.attributes()[0]->type())
+    , label_var_size_(array_schema_.attributes()[0]->var_size())
+    , increasing_labels_(increasing_labels)
+    , index_dim_(array_schema_.domain().dimension_ptr(0))
+    , non_empty_domains_(fragment_metadata_.size())
+    , tile_idx_min_(std::numeric_limits<uint64_t>::max())
+    , tile_idx_max_(std::numeric_limits<uint64_t>::min())
+    , result_tiles_(fragment_metadata_.size())
+    , domain_tile_idx_(fragment_metadata_.size()) {
+  // Sanity checks.
```

Comment:
We have an assert statement in do work that no query conditions are set. If that's necessary, we should probably have a sanity check here too.